zqd/api/bzng.go

```diff
@@ -0,0 +1,84 @@
+package api
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio/bzngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+type BzngSearch struct {
+	reader *bzngio.Reader
+	ctrl   interface{}
+	err    error
+}
+
+func NewBzngSearch(body io.Reader) *BzngSearch {
+	return &BzngSearch{
+		reader: bzngio.NewReader(body, resolver.NewContext()),
+	}
+}
+
+//XXX
+const mtu = 100
+
+func (r *BzngSearch) Pull() (zbuf.Batch, interface{}, error) {
+	v := r.ctrl
+	if v != nil {
+		r.ctrl = nil
+		return nil, v, nil
+	}
+	minTs, maxTs := nano.MaxTs, nano.MinTs
+	var out []*zng.Record
+	for len(out) < mtu {
+		rec, b, err := r.reader.ReadPayload()
+		if err != nil {
+			return nil, nil, err
+		}
+		if rec == nil && b == nil {
+			break
+		}
+		if b != nil {
+			if !bytes.HasPrefix(b, []byte("json:")) {
+				// We expect only json control payloads.
+				// XXX should log error if something else,
+				// but just skip for now.
+				continue
+			}
+			ctrl, err := unpack(b[5:])
+			if err != nil {
+				return nil, nil, err
+			}
+			if end, ok := v.(*TaskEnd); ok {
```

Comment:
It looks to me like `v` will always be nil here because of the check at the beginning of the function?

Reply:
Yep, rewrote this to just implement the `zbuf.Reader` interface which I think is a lot more straightforward.